content/flux/v0.x/function-type-signatures.md

```diff
@@ -0,0 +1,188 @@
+
+Type signatures identify Flux [composite types](/flux/v0.x/data-types/composite/)
+with the following syntaxes:
+
+```js
+[A]             // array type
+[B: A]          // dictionary type
+(param: A) => B // function type
+{}              // record type
+```
+
+### Regular expression types
+```js
+regexp // regular expression type
+```
+
+## Parameter notation
+
+Parameter notation indicates specific behaviors of function parameters.
+
+```js
+?  // Optional parameter
+<- // Pipe receive – indicates the parameter, by default, represents
+   // the piped-forward value or stream of tables
+```
+
+## Type constraints
+
+Some function parameters are "polymorphic" and can support multiple data types.
+Polymorphic parameters are bound by **type constraints**, which define what
+types can be used.
+Type signatures indicate type contraints for specific values using the
+`where A: Constraint` syntax.
+
+For more information about the different type constraints and the types each
+supports, see [Type constraints](/flux/v0.x/spec/types/#type-constraints).
+
+```js
+(v: A) => int where A: Timeable
+```
+
+## Example function type signatures
+
+- [Function without parameters](#function-without-parameters)
+- [Function with parameters](#function-with-parameters)
+- [Pass-through transformation](#pass-through-transformation)
+- [Basic transformation](#basic-transformation)
+- [Transformation that adds a column with an explicit type](#transformation-that-adds-a-column-with-an-explicit-type)
+
+---
+
+#### Function without parameters
+The following type signature describes a function that:
+
+- Has no parameters
+- Returns a time value
+
+```js
+() => time
+```
+
+---
+
+#### Function with parameters
+The following type signature describes a function that:
+
+- Has two parameters:
+  - multiplier _(Optional)_
+  - v ({{< req >}})
+- Returns a value the same type as the two input parameters
+
+```js
+(?multiplier: A, v: A) => A
+```
+
+---
+
+#### Pass-through transformation
+The following type signature describes a
+[transformation](/flux/v0.x/function-types/#transformations) that:
+
+- Takes a stream of tables as piped-forward input
+- Returns the input stream of tables unmodified
+
+```js
+(<-tables: stream[A]) => stream[A]
+```
+
+---
+
+#### Basic transformation
+The following type signature describes a
+[transformation](/flux/v0.x/function-types/#transformations) that:
+
+- Takes a stream of tables as piped-forward input
+- Returns a new, modified stream of tables
+
+```js
+(<-tables: stream[A]) => stream[B]
+```
+
+---
+
+#### Transformation that adds a column with an explicit type
+The following type signature describes a
+[transformation](/flux/v0.x/function-types/#transformations) that:
+
+- Takes a stream of tables as piped-forward input
+- Has a required **tag** parameter
+- Returns a new, modified stream of tables that includes a **tag** column with
+  string values
+
+```js
+(<-tables: stream[A], tag: B) => stream[{A with tag: string}]
```

Comment:
This one confuses me a bit, what is the purpose of the type variable `B`? Should this be instead:


```
(<-tables: stream[A], tag: string) => stream[{A with tag: string}]
```

Meaning why would we allow the `tag` parameter to be any type if we intend to only place a string on the stream data?

Reply:
Sorry, this was the signature from a function that applies `string()` to the `tag` column. It should have a constraint, but it wasn't output when I created it. I'll add the constraint.